# Capstone part 2 - Sentiment Analysis 
Author - Udita Bajaj


In this section, I conduct sentiment analysis. I use the Vader sentiment analyser and then I create a Naive Bayes classifier as a critique of the vader sentiment analyser. 

### Vader Sentiment Intensity Anlayser 
The Vader sentiment analyser assigns a sentiment score to the text after it sums up the intensity of each word present in that text. It computes four scores - negative, neutral, positive and compound.The compound score is computed by normalizing the other three scores. 
I assign the compound score to a new column 'Polarity score', and I create columns for the rest of the scores as they are - 'Negative score', 'Positive score', and 'Neutral score'. The resulting sentiment depends on the value of the polarity score - if a tweet has a Polarity score of less than 0, it is classified as Negative. If a tweet has a Polarity score of more than 0, it is classified as Positive. If a tweet has a Polarity score of 0, it is classified as Neutral. The percentages of classification of tweets is represntated in a pie chart as calculated by the vader sentiment analyser. 

In [80]:
# install vader sentiment
!pip install vaderSentiment

In [81]:
#import sentiment intensity analyser from vader
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

In [82]:
#initialise the Sentiment Intensity analyser, used on the 'tweet_without_stopwords' column
SIA = SentimentIntensityAnalyzer()
newdf["tweet_without_stopwords"]= newdf["tweet_without_stopwords"].astype(str)
#Creating all columns of scores by applying the Sentiment Analyser in a lambda function
newdf['Polarity Score']=newdf["tweet_without_stopwords"].apply(lambda x:SIA.polarity_scores(x)['compound'])


In [83]:
newdf['Neutral Score']=newdf["tweet_without_stopwords"].apply(lambda x:SIA.polarity_scores(x)['neu'])
newdf['Negative Score']=newdf["tweet_without_stopwords"].apply(lambda x:SIA.polarity_scores(x)['neg'])
newdf['Positive Score']=newdf["tweet_without_stopwords"].apply(lambda x:SIA.polarity_scores(x)['pos'])
# Converting 0 to 1 Decimal Score to a Categorical Variable (sentiments) - Positive, Neutral, and Negative
newdf['Sentiment']=''
newdf.loc[newdf['Polarity Score']>0,'Sentiment']='Positive'
newdf.loc[newdf['Polarity Score']==0,'Sentiment']='Neutral'
newdf.loc[newdf['Polarity Score']<0,'Sentiment']='Negative'
#newdf

In [84]:
#create pie chart 
fig_pie = px.pie(newdf, names='Sentiment', title='Tweets Classification', height=250,
                 hole=0.7, color_discrete_sequence=px.colors.qualitative.T10)
fig_pie.update_traces(textfont=dict(color='#fff'))
fig_pie.update_layout(margin=dict(t=50, b=20, l=50, r=25),
                      plot_bgcolor='#2d3035', paper_bgcolor='#2d3035',
                      title_font=dict(size=25, color='#a5a7ab', family="Lato, sans-serif"),
                      font=dict(color='#8a8d93'),
                      legend=dict(orientation="h", yanchor="bottom", y=1, xanchor="right", x=0.8)
                      )

The visualisation depicts that Vader classified 46.4% tweets as positive, 31.8% as negative, and 21.9% as neutral. This is surprising as I expected most tweets about Covid-19 to be negative. Vader might be returning such a result as it may be classifying all tweets with the words 'Booster' and 'new' as positive, whereas their classification depends on the context of the tweet. 

### Manual Investigation 
In this section, I extract 300 tweets (100 each) as classified by Vader. I convert the dataframe to an excel file where I manually classify each of the 300 into the three sentiments.

In [85]:
#list of Neutral sentiment
dfNeutral = newdf[newdf["Sentiment"] == 'Neutral']

In [86]:
#100 random Neutral tweets
sampleNeutral = dfNeutral.sample(n=100)

In [87]:
#list of Positive sentiment
dfPos = newdf[newdf["Sentiment"] == 'Positive']

In [88]:
#100 random Positive tweets
samplePos = dfPos.sample(n=100)

In [89]:
#list of Negative sentiment
dfNeg = newdf[newdf["Sentiment"] == 'Negative']

In [90]:
#100 random Negative tweets
sampleNeg = dfNeg.sample(n=100)

In [91]:
#join all lists
fullsample = sampleNeutral.append(samplePos)
fullSample = fullsample.append(sampleNeg)

In [92]:
#extract relevant columns of dataframe as list, convert to dataframe, and then join
df1 = fullSample['tweet_without_stopwords']
df1 = df1.to_frame()

In [93]:
df2 = fullSample['Sentiment']
df2 = df2.to_frame()

In [94]:
df3 = fullSample['full_text']
df3 = df3.to_frame()

In [95]:
dfsample = df1.join(df2)

In [96]:
dfSample = df3.join(dfsample)

In [97]:
#save as excel file, ready for manual investigation
dfSample.to_excel("dfSampleScraped.xlsx") 

### Importing excel dataset with 'Corrected' sentiment
In this section, I import the manually investigated dataset and calculate how many Vader Sentiment Analyser correctly classified.

In [98]:
import pandas as pd
from pandas import ExcelWriter
from pandas import ExcelFile

In [99]:
dfedited = pd.read_excel("~/Documents/dfSampleScraped2.xlsx")
dfedited

Unnamed: 0                                          full_text  \
0           990  Enjoyment/risk ratios live in my head now. #CO...   
1           904  #SouthKorea to Test #AI-Powered #FacialRecogni...   
2           129  Our 75th KB Friday Recap! \n\nCan you believe ...   
3          1649  So now the experts have decided that you don’t...   
4          1334  Our latest #research on #COVID19, published wi...   
..          ...                                                ...   
295        1732  BREAKING: NY Ethics Committee ordering Andrew ...   
296         135  #COVID19\n@mtosterholm @CovidWatch @ScottGottl...   
297         119  How to show a negative Covid test:\nTake test ...   
298         421  @ARanganathan72 @narendramodi Agree @ARanganat...   
299         919  @miekkk1988 PLEASE get your jab😷\nmy family on...   

                               tweet_without_stopwords Sentiment  Corrected  
0        Enjoyment/risk ratios live head now. #COVID19   Neutral  Negative   
1    #SouthKorea Test #AI-Powered #FacialRecognitio...   Neutral    Neutral  
2    Our 75th KB Friday Recap! Can believe it? Tune...   Neutral    Neutral  
3    So experts decided don’t wait 15mins post covi...   Neutral  Negative   
4    Our latest #research #COVID19, published @Spri...   Neutral    Neutral  
..                                                 ...       ...        ...  
295  BREAKING: NY Ethics Committee ordering Andrew ...  Negative   Negative  
296  #COVID19 @mtosterholm @CovidWatch @ScottGottli...  Negative    Neutral  
297  How show negative Covid test: Take test packet...  Negative   Negative  
298  @ARanganathan72 @narendramodi Agree @ARanganat...  Negative   Negative  
299  @miekkk1988 PLEASE jab😷 family Mums 80th b’day...  Negative   Negative  

[300 rows x 5 columns]

In [100]:
dfSample = fullSample.reset_index()
dfSample

index      screen_name                           name  \
0      158      ncIMPACTsog            ncIMPACT Initiative   
1     1377        Farquetoo                      farquetoo   
2     1539       rantmashup                    Rant Mashup   
3     1593        arcticket   Arcticket ✈️ Visa Assistance   
4      732       CovidData2                     Covid Data   
..     ...              ...                            ...   
295    392          crossky                          Craig   
296   1776  SSweetheartsPod           Slapshot Sweethearts   
297   1362          LaoTzu5  Been There. Done That 🖕🏾 2020   
298   1767      winder_gill                Dr. Winder Gill   
299    689  GirlhoodStudies                GirlhoodStudies   

               location                                        description  \
0       Chapel Hill, NC  Supporting policy innovations through key insi...   
1             australia  live life to the fullest. Drink and eat till y...   
2                                          You've got to be kidding me!...   
3    Manila Philippines  Arcticket is the leading Visa Assistance Agenc...   
4                                   Posting Covid-19 data from web sources   
..                  ...                                                ...   
295           Iowa, USA  Check out my blog! Airline Pilot & Entrepreneu...   
296                      Your go-to for NHL and PHF news and banter hos...   
297         England, UK  🧐 By all means look at this self-declaration. ...   
298     Ontario, Canada  FRCPC Internal Medicine - Clinical Immunology ...   
299                      Interdisciplinary Journal Devoted to the Study...   

     geo_enabled                      created_at  \
0           True  Tue Dec 14 20:36:08 +0000 2021   
1           True  Tue Dec 14 19:21:45 +0000 2021   
2          False  Tue Dec 14 19:11:52 +0000 2021   
3           True  Tue Dec 14 19:08:56 +0000 2021   
4          False  Tue Dec 14 20:00:20 +0000 2021   
..           ...                             ...   
295         True  Tue Dec 14 20:22:13 +0000 2021   
296         True  Tue Dec 14 18:59:31 +0000 2021   
297        False  Tue Dec 14 19:22:41 +0000 2021   
298        False  Tue Dec 14 18:59:56 +0000 2021   
299        False  Tue Dec 14 20:01:47 +0000 2021   

                                             full_text display_text_range  \
0    While local governments work to rebuild, rewor...           [0, 278]   
1    Does her maths add up. #COVID19 #Omicron #vacc...            [0, 62]   
2    To all those working in the #Medical #healthca...           [0, 220]   
3    #Africa, #UK Update: British authorities will ...           [0, 116]   
4    New COVID-19 Data at 2021-12-14 03:00:00 pm ES...            [0, 79]   
..                                                 ...                ...   
295  @downin121 @1ThessCh5 Sara believes she is bei...          [22, 238]   
296  With Craig Smith and Brad Marchand entering #C...           [0, 182]   
297  59,610 new #Covid19 cases today.  \nNot sure h...            [0, 82]   
298  It’s sad to consider cancelling holiday plans ...           [0, 281]   
299  Kids who lost parents to #COVID19 describe 'em...           [0, 121]   

                                                source  \
0    <a href="https://mobile.twitter.com" rel="nofo...   
1    <a href="http://twitter.com/download/android" ...   
2    <a href="http://twitter.com/download/iphone" r...   
3    <a href="http://twitter.com/download/android" ...   
4    <a href="https://kojospace.com/" rel="nofollow...   
..                                                 ...   
295  <a href="http://twitter.com/download/iphone" r...   
296  <a href="http://twitter.com/download/iphone" r...   
297  <a href="https://mobile.twitter.com" rel="nofo...   
298  <a href="http://twitter.com/download/iphone" r...   
299  <a href="https://www.hootsuite.com" rel="nofol...   

                                              entities  \
0    {'hashtags': 

In [101]:
dfedited.drop('Unnamed: 0', axis=1, inplace=True)
#dfedited

In [102]:
#replace all corrected ones with proper punctuation 
dfedited['Corrected'] = dfedited['Corrected'].str.capitalize()

In [103]:
#new column to measure how many match 
dfedited["Suspicion"] = dfedited["Corrected"] != dfedited["Sentiment"]
dfedited

full_text  \
0    Enjoyment/risk ratios live in my head now. #CO...   
1    #SouthKorea to Test #AI-Powered #FacialRecogni...   
2    Our 75th KB Friday Recap! \n\nCan you believe ...   
3    So now the experts have decided that you don’t...   
4    Our latest #research on #COVID19, published wi...   
..                                                 ...   
295  BREAKING: NY Ethics Committee ordering Andrew ...   
296  #COVID19\n@mtosterholm @CovidWatch @ScottGottl...   
297  How to show a negative Covid test:\nTake test ...   
298  @ARanganathan72 @narendramodi Agree @ARanganat...   
299  @miekkk1988 PLEASE get your jab😷\nmy family on...   

                               tweet_without_stopwords Sentiment  Corrected  \
0        Enjoyment/risk ratios live head now. #COVID19   Neutral  Negative    
1    #SouthKorea Test #AI-Powered #FacialRecognitio...   Neutral    Neutral   
2    Our 75th KB Friday Recap! Can believe it? Tune...   Neutral    Neutral   
3    So experts decided don’t wait 15mins post covi...   Neutral  Negative    
4    Our latest #research #COVID19, published @Spri...   Neutral    Neutral   
..                                                 ...       ...        ...   
295  BREAKING: NY Ethics Committee ordering Andrew ...  Negative   Negative   
296  #COVID19 @mtosterholm @CovidWatch @ScottGottli...  Negative    Neutral   
297  How show negative Covid test: Take test packet...  Negative   Negative   
298  @ARanganathan72 @narendramodi Agree @ARanganat...  Negative   Negative   
299  @miekkk1988 PLEASE jab😷 family Mums 80th b’day...  Negative   Negative   

     Suspicion  
0         True  
1        False  
2        False  
3         True  
4        False  
..         ...  
295      False  
296       True  
297      False  
298      False  
299      False  

[300 rows x 5 columns]

In [104]:
#see how many dont match
final = dfedited.loc[dfedited["Suspicion"] == True]
final #107/300 do not match

full_text  \
0    Enjoyment/risk ratios live in my head now. #CO...   
3    So now the experts have decided that you don’t...   
6    BREAKING: #Hawaii sees 214 new #coronavirus ca...   
8    Because a feature of populist democracy is zer...   
10   Delta variant slows recovery in metro Denver o...   
..                                                 ...   
270  @fiannafailparty @MichealMartinTD @RTEOne that...   
273  We are in an information war.\n\nThe crux is t...   
279  Brilliant! Viewer gets in touch with @peter_le...   
290  #COVID19 vaccines teach your T cells to elimin...   
296  #COVID19\n@mtosterholm @CovidWatch @ScottGottl...   

                               tweet_without_stopwords Sentiment  Corrected  \
0        Enjoyment/risk ratios live head now. #COVID19   Neutral  Negative    
3    So experts decided don’t wait 15mins post covi...   Neutral  Negative    
6    BREAKING: #Hawaii sees 214 new #coronavirus ca...   Neutral  Negative    
8    Because feature populist democracy zero attent...   Neutral  Negative    
10   Delta variant slows recovery metro Denver offi...   Neutral  Negative    
..                                                 ...       ...        ...   
270  @fiannafailparty @MichealMartinTD @RTEOne come...  Negative    Neutral   
273  We information war. The crux institutions need...  Negative  Positive    
279  Brilliant! Viewer gets touch @peter_levy he’s ...  Negative  Positive    
290  #COVID19 vaccines teach T cells eliminate infe...  Negative  Positive    
296  #COVID19 @mtosterholm @CovidWatch @ScottGottli...  Negative    Neutral   

     Suspicion  
0         True  
3         True  
6         True  
8         True  
10        True  
..         ...  
270       True  
273       True  
279       True  
290       True  
296       True  

[107 rows x 5 columns]

In [105]:
##DATA CLEANING ##
dfedited["Corrected"].value_counts() 

Negative      94
Positive      67
Neutral       61
Negative      52
Positive      19
Negative       6
Positivr       1
Name: Corrected, dtype: int64

In [106]:
dfedited.loc[dfedited["Corrected"] == "Negative"] 

full_text  \
35   Yup I ordered a pack yesterday. F the province...   
36   Prime Minister Mark Rutte and Health Minister ...   
48   164 New Cases Of #Covid19 Today In Manitoba! T...   
149  @Sillyshib So, the people who wanted #Brexit b...   
150  Couldn't give a fuck hoe many voted for the #M...   
..                                                 ...   
294  Take a look at how condescending the tone is. ...   
295  BREAKING: NY Ethics Committee ordering Andrew ...   
297  How to show a negative Covid test:\nTake test ...   
298  @ARanganathan72 @narendramodi Agree @ARanganat...   
299  @miekkk1988 PLEASE get your jab😷\nmy family on...   

                               tweet_without_stopwords Sentiment Corrected  \
35   Yup I ordered pack yesterday. F provinces maki...   Neutral  Negative   
36   Prime Minister Mark Rutte Health Minister Hugo...   Neutral  Negative   
48   164 New Cases Of #Covid19 Today In Manitoba! T...   Neutral  Negative   
149  @Sillyshib So, people wanted #Brexit believed ...  Positive  Negative   
150  Couldn't give fuck hoe many voted #MaskMandate...  Positive  Negative   
..                                                 ...       ...       ...   
294  Take look condescending tone is. Have poison i...  Negative  Negative   
295  BREAKING: NY Ethics Committee ordering Andrew ...  Negative  Negative   
297  How show negative Covid test: Take test packet...  Negative  Negative   
298  @ARanganathan72 @narendramodi Agree @ARanganat...  Negative  Negative   
299  @miekkk1988 PLEASE jab😷 family Mums 80th b’day...  Negative  Negative   

     Suspicion  
35        True  
36        True  
48        True  
149       True  
150       True  
..         ...  
294      False  
295      False  
297      False  
298      False  
299      False  

[94 rows x 5 columns]

In [107]:
dfedited["Corrected"].replace({"Negative ": "Negative", "Positive ": "Positive"}, inplace=True)

In [108]:
dfedited["Corrected"].value_counts()

Negative      146
Positive       86
Neutral        61
Negative        6
Positivr        1
Name: Corrected, dtype: int64

In [109]:
dfedited.index[dfedited['Corrected']== 'Negative  '].tolist()

[58, 59, 62, 63, 64, 65]

In [110]:
dfedited["Corrected"].replace({"Negative  ": "Negative", "Positivr": "Positive"}, inplace=True)

In [111]:
dfedited["Corrected"].value_counts()

Negative    152
Positive     87
Neutral      61
Name: Corrected, dtype: int64

This sampled and corrected dataset has 152 Negative tweets, 87 Postive tweets, and 61 Neutral tweets. However, since this si to be the training dataset, the tweets need to be balanced in classification. Hence, we must resample and add to this dataset to equalise the number of tweets of each sentiment.

### Re-sampling to acheive balance in sentiment classification 
In this section, I resample ocne more to achieve balance on tweets classified as Negative, Positive, and Neutral. This is to ensure that the Naive Bayes classifier is equally trained on all three sentiments.

In [112]:
rowstodrop = dfSample["index"].to_list()
#rowstodrop

In [113]:
resampledf = newdf.drop(rowstodrop)
#resampledf

In [114]:
#now that we have excluded previous sample, let's resample

In [115]:
dfNeutral2 = resampledf[resampledf["Sentiment"] == 'Neutral']
sampleNeutral2 = dfNeutral2.sample(n=200)

In [116]:
dfPos2 = resampledf[resampledf["Sentiment"] == 'Positive']
samplePos2 = dfPos2.sample(n=200)

In [117]:
fullsample2 = sampleNeutral2.append(samplePos2)
#fullsample2

In [118]:
redf1 = fullsample2['tweet_without_stopwords']
redf1 = redf1.to_frame()

In [119]:
redf2 = fullsample2['Sentiment']
redf2 = redf2.to_frame()

In [120]:
redf3 = fullsample2['full_text']
redf3 = redf3.to_frame()

In [121]:
redfsample = redf1.join(redf2)

In [122]:
redfSample = redf3.join(redfsample)

In [123]:
redfSample.to_excel("redfScraped2.xlsx") 

### Import resampled data 
In this section, I import the resampled data and join it with the previously sampled (and manually classified) dataset.

In [124]:
from pandas import ExcelWriter
from pandas import ExcelFile

In [125]:
redfedit = pd.read_excel("~/Documents/redfScraped2.xlsx")
redfedit

Unnamed: 0                                          full_text  \
0          1499  Wasn't his crew or passengers vaccinated? #cov...   
1           584  Even if #Vaccinated y'all can get #COVID19 ?!?...   
2          1269  Rising above British politics, here’s the view...   
3          1664  Pharmacy-led sites to stay open for longer in ...   
4          1454  Will “B.C” follow suit? #bcpoli #savelives #CO...   
..          ...                                                ...   
395        1482  For scientists - this article implies that the...   
396         986  JUST IN (#coronavirus) via @CNN's @elizabethst...   
397        1830  Boosted #boosted #COVID19 #CovidVaccine https:...   
398         966  NORTHERN CAPE COVID-19 STATISTICS AS AT 14 DEC...   
399        1612  COVID-19 Diagnostics Market\nhttps://t.co/q4fp...   

                               tweet_without_stopwords Sentiment Corrected   
0    Wasn't crew passengers vaccinated? #covid19 ht...   Neutral    Neutral  
1        Even #Vaccinated y'all #COVID19 ?!?!?!?!?!?!?   Neutral   Negative  
2    Rising British politics, here’s view @WHO #Omi...   Neutral    Neutral  
3    Pharmacy-led sites stay open longer COVID-19 j...   Neutral   Positive  
4    Will “B.C” follow suit? #bcpoli #savelives #CO...   Neutral    Neutral  
..                                                 ...       ...        ...  
395  For scientists - article implies Laotian bat c...  Positive        NaN  
396  JUST IN (#coronavirus) via @CNN's @elizabethst...  Positive        NaN  
397  Boosted #boosted #COVID19 #CovidVaccine https:...  Positive        NaN  
398  NORTHERN CAPE COVID-19 STATISTICS AS AT 14 DEC...  Positive        NaN  
399  COVID-19 Diagnostics Market https://t.co/q4fpy...  Positive        NaN  

[400 rows x 5 columns]

In [126]:
redfedit.drop('Unnamed: 0', axis=1, inplace=True)

In [127]:
redfedit["Corrected "].value_counts()

Negative    120
Neutral     106
Positive     89
Name: Corrected , dtype: int64

In [128]:
#create new column indicating whether the vader analysed sentiment an dthe manually corrected sentiment match
redfedit["Suspicion"] = redfedit["Corrected "] != redfedit["Sentiment"]
redfedit

full_text  \
0    Wasn't his crew or passengers vaccinated? #cov...   
1    Even if #Vaccinated y'all can get #COVID19 ?!?...   
2    Rising above British politics, here’s the view...   
3    Pharmacy-led sites to stay open for longer in ...   
4    Will “B.C” follow suit? #bcpoli #savelives #CO...   
..                                                 ...   
395  For scientists - this article implies that the...   
396  JUST IN (#coronavirus) via @CNN's @elizabethst...   
397  Boosted #boosted #COVID19 #CovidVaccine https:...   
398  NORTHERN CAPE COVID-19 STATISTICS AS AT 14 DEC...   
399  COVID-19 Diagnostics Market\nhttps://t.co/q4fp...   

                               tweet_without_stopwords Sentiment Corrected   \
0    Wasn't crew passengers vaccinated? #covid19 ht...   Neutral    Neutral   
1        Even #Vaccinated y'all #COVID19 ?!?!?!?!?!?!?   Neutral   Negative   
2    Rising British politics, here’s view @WHO #Omi...   Neutral    Neutral   
3    Pharmacy-led sites stay open longer COVID-19 j...   Neutral   Positive   
4    Will “B.C” follow suit? #bcpoli #savelives #CO...   Neutral    Neutral   
..                                                 ...       ...        ...   
395  For scientists - article implies Laotian bat c...  Positive        NaN   
396  JUST IN (#coronavirus) via @CNN's @elizabethst...  Positive        NaN   
397  Boosted #boosted #COVID19 #CovidVaccine https:...  Positive        NaN   
398  NORTHERN CAPE COVID-19 STATISTICS AS AT 14 DEC...  Positive        NaN   
399  COVID-19 Diagnostics Market https://t.co/q4fpy...  Positive        NaN   

     Suspicion  
0        False  
1         True  
2        False  
3         True  
4        False  
..         ...  
395       True  
396       True  
397       True  
398       True  
399       True  

[400 rows x 5 columns]

In [129]:
extraneutral = redfedit[redfedit["Corrected "] == 'Neutral']
impneutral = extraneutral[:90]
impneutral

full_text  \
0    Wasn't his crew or passengers vaccinated? #cov...   
2    Rising above British politics, here’s the view...   
4    Will “B.C” follow suit? #bcpoli #savelives #CO...   
5    Leading infectious diseases expert Professor S...   
6    Immunity system to second dose of #COVID19  va...   
..                                                 ...   
185  @CPHO_Canada @JustinTrudeau #CoronaVirus #CoVi...   
186  NEWS: Our Lady's Hospital in Navan says all vi...   
187  Is this the normal you thought of when they to...   
188  I only post a few of these that I see. #Fauci ...   
189  @carlheneghan #COVID19 #Omicron \nCould this b...   

                               tweet_without_stopwords Sentiment Corrected   \
0    Wasn't crew passengers vaccinated? #covid19 ht...   Neutral    Neutral   
2    Rising British politics, here’s view @WHO #Omi...   Neutral    Neutral   
4    Will “B.C” follow suit? #bcpoli #savelives #CO...   Neutral    Neutral   
5    Leading infectious diseases expert Professor S...   Neutral    Neutral   
6    Immunity system second dose #COVID19 vaccinati...   Neutral    Neutral   
..                                                 ...       ...        ...   
185  @CPHO_Canada @JustinTrudeau #CoronaVirus #CoVi...   Neutral    Neutral   
186  NEWS: Our Lady's Hospital Navan says visitors ...   Neutral    Neutral   
187  Is normal thought told ticket back vaccines? #...   Neutral    Neutral   
188  I post I see. #Fauci #EnoughIsEnough #COVID19 ...   Neutral    Neutral   
189  @carlheneghan #COVID19 #Omicron Could looking ...   Neutral    Neutral   

     Suspicion  
0        False  
2        False  
4        False  
5        False  
6        False  
..         ...  
185      False  
186      False  
187      False  
188      False  
189      False  

[90 rows x 5 columns]

In [130]:
extrapos = redfedit[redfedit["Corrected "] == 'Positive']
imppos = extrapos[:63]
imppos

full_text  \
3    Pharmacy-led sites to stay open for longer in ...   
15   City of Saskatoon eyes return of staff to work...   
17   The #Topsfield Board of Health will be hosting...   
27   As of December 13, 2021, 239.9 million people,...   
31   .@VolvoCEGlobal tests fully autonomous, batter...   
..                                                 ...   
247  .@US_FDA authorized Azova #COVID19 Saliva PCR ...   
250  When Profits are so Good, they have to Keep th...   
252  @TVKev Wee don’t want wars either ! But I gues...   
253  How to safely plan a family gathering this hol...   
254  #diagnsotics #COVID19 Love seeing the growth a...   

                               tweet_without_stopwords Sentiment Corrected   \
3    Pharmacy-led sites stay open longer COVID-19 j...   Neutral   Positive   
15   City Saskatoon eyes return staff workplace htt...   Neutral   Positive   
17   The #Topsfield Board Health hosting #COVID19 v...   Neutral   Positive   
27   As December 13, 2021, 239.9 million people, 72...   Neutral   Positive   
31   .@VolvoCEGlobal tests fully autonomous, batter...   Neutral   Positive   
..                                                 ...       ...        ...   
247  .@US_FDA authorized Azova #COVID19 Saliva PCR ...  Positive   Positive   
250  When Profits Good, Keep Up😂 @moderna_tx #COVID...  Positive   Positive   
252  @TVKev Wee don’t want wars either ! But I gues...  Positive   Positive   
253  How safely plan family gathering holiday seaso...  Positive   Positive   
254  #diagnsotics #COVID19 Love seeing growth next ...  Positive   Positive   

     Suspicion  
3         True  
15        True  
17        True  
27        True  
31        True  
..         ...  
247      False  
250      False  
252      False  
253      False  
254      False  

[63 rows x 5 columns]

In [131]:
joined1 = impneutral.append(imppos, sort = False)

In [132]:
joined1 = joined1.rename(columns={"Corrected ": "Corrected"})

In [133]:
dfedited

full_text  \
0    Enjoyment/risk ratios live in my head now. #CO...   
1    #SouthKorea to Test #AI-Powered #FacialRecogni...   
2    Our 75th KB Friday Recap! \n\nCan you believe ...   
3    So now the experts have decided that you don’t...   
4    Our latest #research on #COVID19, published wi...   
..                                                 ...   
295  BREAKING: NY Ethics Committee ordering Andrew ...   
296  #COVID19\n@mtosterholm @CovidWatch @ScottGottl...   
297  How to show a negative Covid test:\nTake test ...   
298  @ARanganathan72 @narendramodi Agree @ARanganat...   
299  @miekkk1988 PLEASE get your jab😷\nmy family on...   

                               tweet_without_stopwords Sentiment Corrected  \
0        Enjoyment/risk ratios live head now. #COVID19   Neutral  Negative   
1    #SouthKorea Test #AI-Powered #FacialRecognitio...   Neutral   Neutral   
2    Our 75th KB Friday Recap! Can believe it? Tune...   Neutral   Neutral   
3    So experts decided don’t wait 15mins post covi...   Neutral  Negative   
4    Our latest #research #COVID19, published @Spri...   Neutral   Neutral   
..                                                 ...       ...       ...   
295  BREAKING: NY Ethics Committee ordering Andrew ...  Negative  Negative   
296  #COVID19 @mtosterholm @CovidWatch @ScottGottli...  Negative   Neutral   
297  How show negative Covid test: Take test packet...  Negative  Negative   
298  @ARanganathan72 @narendramodi Agree @ARanganat...  Negative  Negative   
299  @miekkk1988 PLEASE jab😷 family Mums 80th b’day...  Negative  Negative   

     Suspicion  
0         True  
1        False  
2        False  
3         True  
4        False  
..         ...  
295      False  
296       True  
297      False  
298      False  
299      False  

[300 rows x 5 columns]

In [134]:
#final training dataframe 
myfinal = pd.concat([dfedited, joined1], axis=0) ## why are two corrected columns coming 

In [135]:
#balanced on sentiments 
myfinal["Corrected"].value_counts()

Negative    152
Neutral     151
Positive    150
Name: Corrected, dtype: int64

This dataset is ready to be used in the Naive Bayes Classifer 

### Text Classification with Naive Bayes
In this section, I fit the naive Bayes classifier on the training dataset through sci-kit learn.
Through cross validation, I compute the accuracy of the Naive Bayes classifier.

In [158]:
import sklearn

In [139]:
#turning the column of tweet without stopwords to a list - each row is a tweet
X = myfinal["tweet_without_stopwords"].tolist()
len(X)

453

In [140]:
Y = myfinal["Corrected"].tolist()
len(Y)

453

Here, I vectorize the data to make it into the training dataset. 

In [156]:
from sklearn.feature_extraction.text import CountVectorizer

In [142]:
vectorizer = CountVectorizer()   
X_train = vectorizer.fit_transform(X)
X_train

<453x3634 sparse matrix of type '<class 'numpy.int64'>'
	with 8394 stored elements in Compressed Sparse Row format>

In [143]:
X_train.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

Next I fit the classifier on my dataset (X) and labels (Y)

In [144]:
#use sklearn to initialise classifier 
from sklearn.naive_bayes import MultinomialNB
classifier = MultinomialNB()
classifier

MultinomialNB()

In [145]:
classifier.fit(X_train, Y)

MultinomialNB()

In [146]:
classifier.classes_

array(['Negative', 'Neutral', 'Positive'], dtype='<U8')

In [147]:
# number of instances for each class
classifier.class_count_

array([152., 151., 150.])

In [148]:
# number of features for each class

classifier.feature_count_

array([[0., 2., 0., ..., 1., 2., 1.],
       [4., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.]])

In [149]:
#calculate log probabilities  
classifier.feature_log_prob_

array([[-8.84922702, -7.75061473, -8.84922702, ..., -8.15607984,
        -7.75061473, -8.15607984],
       [-7.10233473, -8.01862547, -8.71177265, ..., -8.71177265,
        -8.71177265, -8.71177265],
       [-8.82453082, -8.82453082, -8.13138364, ..., -8.82453082,
        -8.82453082, -8.82453082]])

Here, I test on one tweet from the training dataset to check if the classifier functions

In [150]:
test1 = X[0]
test1 = [test1]
test1

['Enjoyment/risk ratios live head now. #COVID19']

In [151]:
testVector1 = vectorizer.transform(test1)
testVector1

<1x3634 sparse matrix of type '<class 'numpy.int64'>'
	with 7 stored elements in Compressed Sparse Row format>

In [152]:
testVector1.toarray()

array([[0, 0, 0, ..., 0, 0, 0]])

In [153]:
classifier.predict(testVector1)

array(['Negative'], dtype='<U8')

In [163]:
from sklearn.model_selection import cross_validate
from sklearn.model_selection import cross_val_predict

The classifier predicted the sentiment of the test vector correctly. This indicates that the classifier works.

Next, I perform cross-validation to calcuate the scores of the model. I use cross-validation to predict the sentiments of all the tweets and then to cacluate the average accuracy of the model.

In [161]:
# Perform cross-validation
cv_results_cv = cross_validate(classifier, 
                                X_train,
                                Y,
                                cv=5
                                )
# Show results
cv_results_cv   #three classes so 0.3 is threshold 

{'fit_time': array([0.00552797, 0.00194907, 0.00251102, 0.00177813, 0.00166082]),
 'score_time': array([0.00097799, 0.0004189 , 0.00062799, 0.00041604, 0.00039411]),
 'test_score': array([0.58241758, 0.49450549, 0.42857143, 0.52222222, 0.58888889])}

In [164]:
cv_results_cvp = cross_val_predict(classifier, 
                                   X_train,
                                   Y,
                                   cv=5
                                   )
# Show results
cv_results_cvp

array(['Negative', 'Neutral', 'Neutral', 'Negative', 'Positive',
       'Neutral', 'Neutral', 'Neutral', 'Negative', 'Negative', 'Neutral',
       'Positive', 'Positive', 'Neutral', 'Neutral', 'Positive',
       'Neutral', 'Neutral', 'Neutral', 'Negative', 'Positive',
       'Negative', 'Negative', 'Negative', 'Negative', 'Positive',
       'Neutral', 'Neutral', 'Neutral', 'Positive', 'Positive', 'Neutral',
       'Positive', 'Negative', 'Negative', 'Positive', 'Neutral',
       'Neutral', 'Positive', 'Neutral', 'Negative', 'Neutral', 'Neutral',
       'Neutral', 'Neutral', 'Negative', 'Neutral', 'Neutral', 'Neutral',
       'Positive', 'Neutral', 'Positive', 'Neutral', 'Neutral', 'Neutral',
       'Neutral', 'Neutral', 'Negative', 'Positive', 'Neutral',
       'Negative', 'Neutral', 'Neutral', 'Neutral', 'Negative',
       'Negative', 'Neutral', 'Neutral', 'Negative', 'Positive',
       'Neutral', 'Negative', 'Positive', 'Neutral', 'Neutral',
       'Positive', 'Neutral', 'Negative', 

In [165]:
#calcuate the avarage accuracy of the model 
average_accuracy = sum(cv_results_cv['test_score'])/len(cv_results_cv['test_score'])
average_accuracy 
#0.5233211233211235

0.5233211233211235

Our classifer has an average accuracy of 0.533. Since this a multinomial 3 category classifier, the threshold value for the model is 0.33 (⅓). A value of 0.33 would suggest that the model works on pure randomization, and so, a value of 0.5233 suggests that the model works well. 

### Predicting sentiments for 2000 tweets 
In this section, I use the Naive Bayes classifer, now trained on sentiment classification, to predict the sentiments of all 2000 tweets. I represent the predictions in a pie chart. 

In [166]:
X1 = newdf["tweet_without_stopwords"].tolist()
len(X1)

2000

In [167]:
testVectorfull = vectorizer.transform(X1)
testVectorfull

<2000x3634 sparse matrix of type '<class 'numpy.int64'>'
	with 29906 stored elements in Compressed Sparse Row format>

In [168]:
testVectorfull.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 1, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [169]:
myPredict = classifier.predict(testVectorfull)

In [170]:
type(myPredict)

numpy.ndarray

In [171]:
NaiveB = pd.DataFrame(myPredict, columns = ["Naive Bayes"])
NaiveB

Naive Bayes
0       Negative
1       Negative
2       Positive
3       Negative
4       Negative
...          ...
1995    Negative
1996    Negative
1997    Positive
1998    Negative
1999    Negative

[2000 rows x 1 columns]

In [172]:
fulldf = newdf.join(NaiveB)
#fulldf

In [176]:
fulldf["Suspicion"] = fulldf["Sentiment"] != fulldf["Naive Bayes"]
fulldf

screen_name                            name  \
0      HollyGreyhound    Bonnie & 🌈Holly Greyhound🌈💔😭   
1            PutnamDV                DV Putnam County   
2           VetCanuck                  Terence Graham   
3      DVOrangeCounty                DV Orange County   
4         DailyNassau       Daily Voice Nassau County   
...               ...                             ...   
1995  SpiritualCrypt1        Alpha-Wizz the #NFT Pro🅕   
1996    InfectDisNews         Infectious Disease News   
1997    UHANDPartners                           UHAND   
1998          Faal_26               Mohamed Falih Ali   
1999       beajayemac  bruce macfarlane #RejoinEU A49   

                            location  \
0                    Nottingham U.K.   
1                  Putnam County, NY   
2                             Canada   
3                                      
4                                      
...                              ...   
1995  Tweets are not Financialadvice   
1996                   Thorofare, NJ   
1997                     Houston, TX   
1998                        Maldives   
1999                         Britain   

                                            description  geo_enabled  \
0     A sweetheart who loved cuddles OTRB 21/6/20💔Li...        False   
1     Daily Voice Putnam County covers our friends a...        False   
2     #Veterans  #MotherNature 🇨🇦❤⚽️👀\n#ClimateEmerg...         True   
3                                                              False   
4                                                              False   
...                                                 ...          ...   
1995  Dad and husband\nPrometheus of Fantom! \n\nI c...        False   
1996            The content you need, when you need it.        False   
1997  @UHouston & @MDAndersonNews collaboration desi...        False   
1998            Soldier, MBBS, MD Pulmonology, 🇲🇻 first        False   
1999  Retired Health Physicist. Space Time Traveller...        False   

                          created_at  \
0     Tue Dec 14 20:46:38 +0000 2021   
1     Tue Dec 14 20:46:38 +0000 2021   
2     Tue Dec 14 20:46:37 +0000 2021   
3     Tue Dec 14 20:46:37 +0000 2021   
4     Tue Dec 14 20:46:37 +0000 2021   
...                              ...   
1995  Tue Dec 14 18:45:33 +0000 2021   
1996  Tue Dec 14 18:45:30 +0000 2021   
1997  Tue Dec 14 18:45:24 +0000 2021   
1998  Tue Dec 14 18:45:22 +0000 2021   
1999  Tue Dec 14 18:45:21 +0000 2021   

                                              full_text display_text_range  \
0     Just how many of your constituents have had th...           [0, 190]   
1     * Cuomo "American Crisis" * An ethics committe...           [0, 219]   
2     @joncoopertweets I'm a vet who had the anthrax...          [17, 144]   
3     * Cuomo "American Crisis" * An ethics committe...           [0, 219]   
4     * Cuomo "American Crisis" * An ethics committe...           [0, 219]   
...                                                 ...                ...   
1995  I think, people are about to wake up from thei...           [0, 279]   
1996  A final analysis confirmed that Pfizer’s inves...           [0, 253]   
1997  We are pleased to announce that our paper deta...           [0, 227]   
1998  An analysis by the largest private health insu...           [0, 273]   
1999  This is the ratio of the number people admitte...           [0, 213]   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://www.dailyvoice.com" rel="nofol...   
2     <a href="http://twitter.com/download/android" ...   
3     <a href="http://www.dailyvoice.com" rel="nofol...   
4     <a href="https://dailyvoice.com/" rel="nofollo...   
...                                                 ...   
1995  <a href="https://mobile.twitter.com" rel="nofo...   
1996  <a href="https://www.healio.com" rel="nofollow...   
1997  <a href="https://mobile.

In [180]:
#save fulldf to excel and inspect missing values for location 
#for this capstone round submission - use what you've got and then shift to choropeth maps 
#relevant columns for choropeth map are sentiment and location 

In [181]:
fulldf.to_excel('fulldftouse.xlsx') #saving full dataframe to manually inspect missing values for location

In [178]:
print(fulldf["Naive Bayes"].value_counts())

Negative    812
Positive    731
Neutral     457
Name: Naive Bayes, dtype: int64


In [179]:
print(fulldf["location"].value_counts())

                           442
United Kingdom              29
New York, NY                23
Whitby, Ontario, Canada     21
United States               19
                          ... 
Harare, Zimbabwe             1
Brisbane, Queensland         1
Elk Grove, CA                1
Manchester England           1
Barnstaple, England          1
Name: location, Length: 927, dtype: int64


In [182]:
fig_pie = px.pie(fulldf, names='Naive Bayes', title='Tweets Classifictaion', height=250,
                 hole=0.7, color_discrete_sequence=px.colors.qualitative.T10)
fig_pie.update_traces(textfont=dict(color='#fff'))
fig_pie.update_layout(margin=dict(t=80, b=30, l=70, r=40),
                      plot_bgcolor='#2d3035', paper_bgcolor='#2d3035',
                      title_font=dict(size=25, color='#a5a7ab', family="Lato, sans-serif"),
                      font=dict(color='#8a8d93'),
                      legend=dict(orientation="h", yanchor="bottom", y=1, xanchor="right", x=0.8)
                      )

It can be seen that in contrast to the Vader classification which listed most tweets as positive, the Naive Bayes classifier classifies 41.5% tweets as Negative, 35% as positive, and 23.5% as neutral. We see a difference in the classifiers. This highlights a growing need for classifiers, especially sentiment analysis classifiers, to be developed by taking into account social context in order to analyse people's opinions, as well as growing inequalities that social phenomena such as Covid-19 bring to light.

#### In this next step, we input the dataframe with the tweets that were manually classified according to location

In [196]:
mydf = pd.read_excel("fulldftouse.xlsx")

In [197]:
mydf

Unnamed: 0      screen_name                            name  \
0              0   HollyGreyhound    Bonnie & 🌈Holly Greyhound🌈💔😭   
1              1         PutnamDV                DV Putnam County   
2              2        VetCanuck                  Terence Graham   
3              3   DVOrangeCounty                DV Orange County   
4              4      DailyNassau       Daily Voice Nassau County   
...          ...              ...                             ...   
1995        1995  SpiritualCrypt1        Alpha-Wizz the #NFT Pro🅕   
1996        1996    InfectDisNews         Infectious Disease News   
1997        1997    UHANDPartners                           UHAND   
1998        1998          Faal_26               Mohamed Falih Ali   
1999        1999       beajayemac  bruce macfarlane #RejoinEU A49   

                            location          actual  \
0                    Nottingham U.K.  United Kingdom   
1                  Putnam County, NY             USA   
2                             Canada          Canada   
3                                NaN             USA   
4                                NaN             USA   
...                              ...             ...   
1995  Tweets are not Financialadvice             NaN   
1996                   Thorofare, NJ             NaN   
1997                     Houston, TX             NaN   
1998                        Maldives             NaN   
1999                         Britain             NaN   

                                            description  geo_enabled  \
0     A sweetheart who loved cuddles OTRB 21/6/20💔Li...        False   
1     Daily Voice Putnam County covers our friends a...        False   
2     #Veterans  #MotherNature 🇨🇦❤⚽️👀\n#ClimateEmerg...         True   
3                                                   NaN        False   
4                                                   NaN        False   
...                                                 ...          ...   
1995  Dad and husband\nPrometheus of Fantom! \n\nI c...        False   
1996            The content you need, when you need it.        False   
1997  @UHouston & @MDAndersonNews collaboration desi...        False   
1998            Soldier, MBBS, MD Pulmonology, 🇲🇻 first        False   
1999  Retired Health Physicist. Space Time Traveller...        False   

                          created_at  \
0     Tue Dec 14 20:46:38 +0000 2021   
1     Tue Dec 14 20:46:38 +0000 2021   
2     Tue Dec 14 20:46:37 +0000 2021   
3     Tue Dec 14 20:46:37 +0000 2021   
4     Tue Dec 14 20:46:37 +0000 2021   
...                              ...   
1995  Tue Dec 14 18:45:33 +0000 2021   
1996  Tue Dec 14 18:45:30 +0000 2021   
1997  Tue Dec 14 18:45:24 +0000 2021   
1998  Tue Dec 14 18:45:22 +0000 2021   
1999  Tue Dec 14 18:45:21 +0000 2021   

                                              full_text display_text_range  \
0     Just how many of your constituents have had th...           [0, 190]   
1     * Cuomo "American Crisis" * An ethics committe...           [0, 219]   
2     @joncoopertweets I'm a vet who had the anthrax...          [17, 144]   
3     * Cuomo "American Crisis" * An ethics committe...           [0, 219]   
4     * Cuomo "American Crisis" * An ethics committe...           [0, 219]   
...                                                 ...                ...   
1995  I think, people are about to wake up from thei...           [0, 279]   
1996  A final analysis confirmed that Pfizer’s inves...           [0, 253]   
1997  We are pleased to announce that our paper deta...           [0, 227]   
1998  An analysis by the largest private health insu...           [0, 273]   
1999  This is the ratio of the number people admitte...           [0, 213]   

      ...                                           entities  \
0     ...  {'hashtags': [{'text': 'ToriesPartiedWhilePeop...   
1     ...  {'hashtags': [{'text': 'COVID19', 'indices': [...   
2     ...  {'hashtags': [{'te

In [199]:
mydf["actual"].value_counts()

USA                      295
United Kingdom           127
Canada                   125
Australia                 29
USA                       15
Germany                   11
Wales                      7
South Africa               7
India                      6
Spain                      6
Kenya                      5
New Zealand                3
Switzerland                3
Switzerland                3
Canada                     3
Trinidad and Tobago        2
Italy                      2
Iran                       2
France                     2
Pakistan                   2
United Arab Emirates       2
France                     2
Netherlands                2
Uganda                     2
South America              2
Ireland                    2
Ireland                    2
Cuba                       1
Cape Town                  1
Dubai                      1
Jamaica                    1
Africa/Nairobi             1
Somalia                    1
Ghana                      1
Austria       

In [201]:
mydf.index[mydf['actual']== 'USA '].tolist() #data cleaning 

[7, 21, 25, 27, 29, 30, 33, 37, 46, 54, 66, 74, 75, 91, 105]

In [204]:
mydf.index[mydf['actual']== 'Canada '].tolist() #data cleaning

[144, 197, 266]

In [206]:
mydf.index[mydf['actual']== 'Ontaria'].tolist() #data cleaning 

[976]

In [207]:
mydf["actual"].replace({"USA ": "USA", "Canada ": "Canada", "Ontaria" : "Canada"}, inplace=True) #data cleaning 

In [208]:
mydf["actual"].value_counts()

USA                      310
Canada                   129
United Kingdom           127
Australia                 29
Germany                   11
Wales                      7
South Africa               7
Spain                      6
India                      6
Kenya                      5
New Zealand                3
Switzerland                3
Switzerland                3
Trinidad and Tobago        2
Italy                      2
Iran                       2
France                     2
Pakistan                   2
France                     2
Netherlands                2
Uganda                     2
South America              2
Ireland                    2
United Arab Emirates       2
Ireland                    2
Cuba                       1
Cape Town                  1
Dubai                      1
Jamaica                    1
Israel                     1
Somalia                    1
Ghana                      1
Africa/Nairobi             1
Austria                    1
Ethiopia      

In [216]:
#Choosing to do only do USA, United Kingdom, and Canada for now
mydfUSA = mydf[mydf["actual"] == "USA"]
mydfUSA

Unnamed: 0      screen_name                                   name  \
1              1         PutnamDV                       DV Putnam County   
3              3   DVOrangeCounty                       DV Orange County   
4              4      DailyNassau              Daily Voice Nassau County   
5              5    Dayna_Fisher1                           Dayna Fisher   
7              7       JennaForMO    Jenna Roberson, State Senate MO-SD2   
...          ...              ...                                    ...   
998          998     9erchic4life                           seasoneddiva   
1000        1000    chitownrapper          AL Rapp AKA Allosaurus Raptor   
1002        1002        tomleykis  Even MORE Fully Vaccinated Tom Leykis   
1004        1004     Ken_DowneyJr                         Ken Downey Jr.   
1006        1006  ImariScarbrough                       Imari Scarbrough   

                           location actual  \
1                 Putnam County, NY    USA   
3                               NaN    USA   
4                               NaN    USA   
5                               USA    USA   
7                    Wentzville, MO    USA   
...                             ...    ...   
998                            Cali    USA   
1000           Chicago IL / Taos NM    USA   
1002  Santa Barbara Cty, California    USA   
1004                  Thorofare, NJ    USA   
1006                  United States    USA   

                                            description  geo_enabled  \
1     Daily Voice Putnam County covers our friends a...        False   
3                                                   NaN        False   
4                                                   NaN        False   
5     Health, MS, RYT 200, patient-centered marcomms...        False   
7     Dem Candidate for MO Senate District 2. Tired ...        False   
...                                                 ...          ...   
998              you do better...when you know better 😉         True   
1000  Eats meat in moderation\nIndependent with Prog...        False   
1002  Hear our podcast at https://t.co/LC270QtHNj. B...        False   
1004  I cover infectious diseases @InfectDisNews 🦠💉M...         True   
1006  Freelance journalist | Open for assignments | ...        False   

                          created_at  \
1     Tue Dec 14 20:46:38 +0000 2021   
3     Tue Dec 14 20:46:37 +0000 2021   
4     Tue Dec 14 20:46:37 +0000 2021   
5     Tue Dec 14 20:46:36 +0000 2021   
7     Tue Dec 14 20:46:27 +0000 2021   
...                              ...   
998   Tue Dec 14 19:44:53 +0000 2021   
1000  Tue Dec 14 19:44:48 +0000 2021   
1002  Tue Dec 14 19:44:36 +0000 2021   
1004  Tue Dec 14 19:44:29 +0000 2021   
1006  Tue Dec 14 19:44:28 +0000 2021   

                                              full_text display_text_range  \
1     * Cuomo "American Crisis" * An ethics committe...           [0, 219]   
3     * Cuomo "American Crisis" * An ethics committe...           [0, 219]   
4     * Cuomo "American Crisis" * An ethics committe...           [0, 219]   
5     @GovTimWalz @PennyWheelerMD How about making C...          [28, 124]   
7     How does #COVID19 affect us?\n\nMy partner's d...           [0, 264]   
...                                                 ...                ...   
998   Bravo to @TishJames for advocating for HOME HE...           [0, 254]   
1000  This is dead wrong.....this is how the US star...           [0, 272]   
1002  You thought I was crazy when I predicted what'...            [0, 92]   
1004  @Tim6992 @EliseStefanik @morganfmckay TX #COVI...          [38, 204]   
1006  @ncdhhs officials are urging #NorthCarolina re...           [0, 179]   

      ...                                           entities  \
1     ...  {'hashtags': [{'text': 'COVID19', 'indices': [...   
3     ...  {'hashtags': [{'text': 'COVID19', 'indices': [...   
4     ...  {'hashtags': [{'text': 'COVID19', 'indices': [...   
5

In [228]:
#rename USA to United States of America (as in the geojson) and assign geographical identifier
mydfUSA["actual"].replace({"USA": "United States of America"}, inplace=True)
mydfUSA["id"] = "USA"    

/Users/udita/opt/anaconda3/lib/python3.8/site-packages/pandas/core/series.py:4509: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

<ipython-input-228-d3e174888a16>:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [244]:
mydfCanada = mydf[mydf["actual"] == "Canada"]
mydfCanada

Unnamed: 0      screen_name                    name  \
2              2        VetCanuck          Terence Graham   
35            35          rohin5r  Rohin Minocha-Mckenney   
36            36         trycom88                Mr Sacha   
47            47       Muskoka411         Muskoka411 News   
48            48   MiltonReporter          MiltonReporter   
...          ...              ...                     ...   
976          976  ChrisLewisEssex    Chris Lewis MP Essex   
979          979        OtagoGrad               Brad Rush   
990          990  NicolasRobidoux        Nicolas Robidoux   
999          999  Darrell_Samuels         Darrell Samuels   
1010        1010       drjeff_unb      Jeffrey J. McNally   

                        location  actual  \
2                         Canada  Canada   
35                        Canada  Canada   
36               Alberta, Canada  Canada   
47              Muskoka, Ontario  Canada   
48               Milton, Ontario  Canada   
...                          ...     ...   
976               Essex, Ontario  Canada   
979                  Calgary, AB  Canada   
990                     Montréal  Canada   
999              Mississauga, ON  Canada   
1010  Fredericton, New Brunswick  Canada   

                                            description  geo_enabled  \
2     #Veterans  #MotherNature 🇨🇦❤⚽️👀\n#ClimateEmerg...         True   
35    Mount Allison University | Biology and Commerc...         True   
36                         Teacher! All over the place!        False   
47    News organization providing the best in breaki...         True   
48    Bringing a piece of #MiltonON to the world, an...        False   
...                                                 ...          ...   
976   Chris Lewis: CPC Member of Parliament (FEDERAL...         True   
979   ClimateCrisis skeptic. EX: biz consultant (org...         True   
990   Thousandth monkey\nil/lui\nThe early bird gets...        False   
999   Always-learning believer, mental health surviv...        False   
1010  Dad, business professor, newshound, & avid wor...         True   

                          created_at  \
2     Tue Dec 14 20:46:37 +0000 2021   
35    Tue Dec 14 20:44:30 +0000 2021   
36    Tue Dec 14 20:44:30 +0000 2021   
47    Tue Dec 14 20:43:46 +0000 2021   
48    Tue Dec 14 20:43:43 +0000 2021   
...                              ...   
976   Tue Dec 14 19:46:37 +0000 2021   
979   Tue Dec 14 19:46:23 +0000 2021   
990   Tue Dec 14 19:45:05 +0000 2021   
999   Tue Dec 14 19:44:52 +0000 2021   
1010  Tue Dec 14 19:44:13 +0000 2021   

                                              full_text display_text_range  \
2     @joncoopertweets I'm a vet who had the anthrax...          [17, 144]   
35    I’m really glad to see quick action on part of...           [0, 250]   
36    @JustinTrudeau Hey, when will you renew CERB (...           [0, 207]   
47    Effective immediately, all visitors must be fu...           [0, 182]   
48    Meanwhile, new protocols for Long-Term Care ha...           [0, 205]   
...                                                 ...                ...   
976   Just moments ago I stood in the House of Commo...           [0, 274]   
979   Disease Mitigation Measures in the Control of ...           [0, 266]   
990   Enjoyment/risk ratios live in my head now. #CO...            [0, 51]   
999   Remind me again.... Is the light at the end of...            [0, 77]   
1010  Pfizer confirms COVID-19 pill's results, poten...            [0, 97]   

      ...                                           entities  \
2     ...  {'hashtags': [{'text': 'COVID19', 'indices': [...   
35    ...  {'hashtags': [{'text': 'COVID19', 'indices': [...   
36    ...  {'hashtags': [{'text': 'COVID19', 'indices': [...   
47    ...  {'hashtags': [{'text': 'News', 'indices': [159...   
48    ...  {'hashtags': [{'text': 'onpoli', 'indices': [1...   
...   ...                                                ...   
976   ...  {'

In [246]:
#assign Canada the geographical identifier 
mydfCanada["id"] = "CAN"
mydfCanada

<ipython-input-246-08aafe3e471f>:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Unnamed: 0      screen_name                    name  \
2              2        VetCanuck          Terence Graham   
35            35          rohin5r  Rohin Minocha-Mckenney   
36            36         trycom88                Mr Sacha   
47            47       Muskoka411         Muskoka411 News   
48            48   MiltonReporter          MiltonReporter   
...          ...              ...                     ...   
976          976  ChrisLewisEssex    Chris Lewis MP Essex   
979          979        OtagoGrad               Brad Rush   
990          990  NicolasRobidoux        Nicolas Robidoux   
999          999  Darrell_Samuels         Darrell Samuels   
1010        1010       drjeff_unb      Jeffrey J. McNally   

                        location  actual  \
2                         Canada  Canada   
35                        Canada  Canada   
36               Alberta, Canada  Canada   
47              Muskoka, Ontario  Canada   
48               Milton, Ontario  Canada   
...                          ...     ...   
976               Essex, Ontario  Canada   
979                  Calgary, AB  Canada   
990                     Montréal  Canada   
999              Mississauga, ON  Canada   
1010  Fredericton, New Brunswick  Canada   

                                            description  geo_enabled  \
2     #Veterans  #MotherNature 🇨🇦❤⚽️👀\n#ClimateEmerg...         True   
35    Mount Allison University | Biology and Commerc...         True   
36                         Teacher! All over the place!        False   
47    News organization providing the best in breaki...         True   
48    Bringing a piece of #MiltonON to the world, an...        False   
...                                                 ...          ...   
976   Chris Lewis: CPC Member of Parliament (FEDERAL...         True   
979   ClimateCrisis skeptic. EX: biz consultant (org...         True   
990   Thousandth monkey\nil/lui\nThe early bird gets...        False   
999   Always-learning believer, mental health surviv...        False   
1010  Dad, business professor, newshound, & avid wor...         True   

                          created_at  \
2     Tue Dec 14 20:46:37 +0000 2021   
35    Tue Dec 14 20:44:30 +0000 2021   
36    Tue Dec 14 20:44:30 +0000 2021   
47    Tue Dec 14 20:43:46 +0000 2021   
48    Tue Dec 14 20:43:43 +0000 2021   
...                              ...   
976   Tue Dec 14 19:46:37 +0000 2021   
979   Tue Dec 14 19:46:23 +0000 2021   
990   Tue Dec 14 19:45:05 +0000 2021   
999   Tue Dec 14 19:44:52 +0000 2021   
1010  Tue Dec 14 19:44:13 +0000 2021   

                                              full_text display_text_range  \
2     @joncoopertweets I'm a vet who had the anthrax...          [17, 144]   
35    I’m really glad to see quick action on part of...           [0, 250]   
36    @JustinTrudeau Hey, when will you renew CERB (...           [0, 207]   
47    Effective immediately, all visitors must be fu...           [0, 182]   
48    Meanwhile, new protocols for Long-Term Care ha...           [0, 205]   
...                                                 ...                ...   
976   Just moments ago I stood in the House of Commo...           [0, 274]   
979   Disease Mitigation Measures in the Control of ...           [0, 266]   
990   Enjoyment/risk ratios live in my head now. #CO...            [0, 51]   
999   Remind me again.... Is the light at the end of...            [0, 77]   
1010  Pfizer confirms COVID-19 pill's results, poten...            [0, 97]   

      ...                                               user  \
2     ...  {'id': 809384904467316736, 'id_str': '80938490...   
35    ...  {'id': 2723702133, 'id_str': '2723702133', 'na...   
36    ...  {'id': 76865074, 'id_str': '76865074', 'name':...   
47    ...  {'id': 380680185, 'id_str': '380680185', 'name...   
48    ...  {'id': 1055242933098684417, 'id_str': '1055242...   
...   ...                                                ...   
976   ...  {'

In [247]:
mydfUK = mydf[mydf["actual"] == "United Kingdom"]
mydfUK

Unnamed: 0      screen_name                          name  \
0              0   HollyGreyhound  Bonnie & 🌈Holly Greyhound🌈💔😭   
6              6           brynll                Bryn Llewellyn   
11            11   martingfindlay                Martin Findlay   
17            17  msleeplessagain                             💪   
18            18       thehrbooth                  The HR Booth   
...          ...              ...                           ...   
975          975         SiDedman                  Simon Dedman   
985          985     lottyleeming             Charlotte Leeming   
992          992     nanakumi1976            David Ofosu-Appiah   
1007        1007     ahmedaftab68            Sakkaf Ahmed Aftab   
1009        1009      clivewelsh5                   Clive Welsh   

                        location          actual  \
0                Nottingham U.K.  United Kingdom   
6        West Yorkshire & London  United Kingdom   
11        Aberdeen, Scotland, UK  United Kingdom   
17       England, United Kingdom  United Kingdom   
18         Dunfermline, Scotland  United Kingdom   
...                          ...             ...   
975            Chelmsford-London  United Kingdom   
985   Yorkshire/Northern England  United Kingdom   
992      England, United Kingdom  United Kingdom   
1007             Elsham, England  United Kingdom   
1009            Bexhill, England  United Kingdom   

                                            description  geo_enabled  \
0     A sweetheart who loved cuddles OTRB 21/6/20💔Li...        False   
6     #moveandlearn advocate * FRSA * @HundrEDorg am...         True   
11    had two Covid-19 doses, a Covid-19 booster and...         True   
17    “The cry of the poor is not always just, but i...         True   
18    HR Consultancy, providing solutions for SME bu...        False   
...                                                 ...          ...   
975   Political Reporter @BBCNews @BBCEssex former @...         True   
985   Award-winning Senior Broadcast Journalist & TV...         True   
992                             Law Advocate,Consultant         True   
1007  Eye doc,Chair BMA Yorkshire Consultant Committ...        False   
1009                         Retired from all but golf.        False   

                          created_at  \
0     Tue Dec 14 20:46:38 +0000 2021   
6     Tue Dec 14 20:46:28 +0000 2021   
11    Tue Dec 14 20:46:10 +0000 2021   
17    Tue Dec 14 20:45:40 +0000 2021   
18    Tue Dec 14 20:45:39 +0000 2021   
...                              ...   
975   Tue Dec 14 19:46:48 +0000 2021   
985   Tue Dec 14 19:45:29 +0000 2021   
992   Tue Dec 14 19:45:00 +0000 2021   
1007  Tue Dec 14 19:44:20 +0000 2021   
1009  Tue Dec 14 19:44:16 +0000 2021   

                                              full_text display_text_range  \
0     Just how many of your constituents have had th...           [0, 190]   
6     How low can scammers go?\n#COVID19 #scam https...            [0, 39]   
11    Good effort, 45.6K #Scotland #booster jabs Mon...           [0, 280]   
17    We live in a society where you have to show a ...           [0, 213]   
18    The First Minister made an announcement to the...           [0, 274]   
...                                                 ...                ...   
975   96 Conservative MPs voting against the governm...           [0, 253]   
985   Tory rebellion doesn’t stop covid passes becom...            [0, 73]   
992   Boris Johnson suffers rebellion by almost 100 ...           [0, 282]   
1007  @DRTomlinsonEP @DrAnneMurphy @CareQualityComm ...         [140, 328]   
1009  @LukeHurrenpfc I thought I was replying to a g...          [15, 293]   

      ...                                           entities  \
0     ...  {'hashtags': [{'text': 'ToriesPartiedWhilePeop...   
6     ...  {'hashtags': [{'text': 'COVID19', 'indices': [...   
11    ...  {'hashtags': [{'text': 'Scotland', 'indices': ...   
17    ...  {'hashtags': [{'text': 'Om

In [248]:
#assign geographical identifier 
mydfUK["id"] = "GBR"

<ipython-input-248-8592ba648fcd>:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [255]:
#combine dataframes 
finaldf = pd.concat([mydfUSA, mydfCanada], axis=0)
finaldf2 = pd.concat([finaldf, mydfUK], axis = 0)
finaldf2

Unnamed: 0     screen_name                                 name  \
1              1        PutnamDV                     DV Putnam County   
3              3  DVOrangeCounty                     DV Orange County   
4              4     DailyNassau            Daily Voice Nassau County   
5              5   Dayna_Fisher1                         Dayna Fisher   
7              7      JennaForMO  Jenna Roberson, State Senate MO-SD2   
...          ...             ...                                  ...   
975          975        SiDedman                         Simon Dedman   
985          985    lottyleeming                    Charlotte Leeming   
992          992    nanakumi1976                   David Ofosu-Appiah   
1007        1007    ahmedaftab68                   Sakkaf Ahmed Aftab   
1009        1009     clivewelsh5                          Clive Welsh   

                        location                    actual  \
1              Putnam County, NY  United States of America   
3                            NaN  United States of America   
4                            NaN  United States of America   
5                            USA  United States of America   
7                 Wentzville, MO  United States of America   
...                          ...                       ...   
975            Chelmsford-London            United Kingdom   
985   Yorkshire/Northern England            United Kingdom   
992      England, United Kingdom            United Kingdom   
1007             Elsham, England            United Kingdom   
1009            Bexhill, England            United Kingdom   

                                            description  geo_enabled  \
1     Daily Voice Putnam County covers our friends a...        False   
3                                                   NaN        False   
4                                                   NaN        False   
5     Health, MS, RYT 200, patient-centered marcomms...        False   
7     Dem Candidate for MO Senate District 2. Tired ...        False   
...                                                 ...          ...   
975   Political Reporter @BBCNews @BBCEssex former @...         True   
985   Award-winning Senior Broadcast Journalist & TV...         True   
992                             Law Advocate,Consultant         True   
1007  Eye doc,Chair BMA Yorkshire Consultant Committ...        False   
1009                         Retired from all but golf.        False   

                          created_at  \
1     Tue Dec 14 20:46:38 +0000 2021   
3     Tue Dec 14 20:46:37 +0000 2021   
4     Tue Dec 14 20:46:37 +0000 2021   
5     Tue Dec 14 20:46:36 +0000 2021   
7     Tue Dec 14 20:46:27 +0000 2021   
...                              ...   
975   Tue Dec 14 19:46:48 +0000 2021   
985   Tue Dec 14 19:45:29 +0000 2021   
992   Tue Dec 14 19:45:00 +0000 2021   
1007  Tue Dec 14 19:44:20 +0000 2021   
1009  Tue Dec 14 19:44:16 +0000 2021   

                                              full_text display_text_range  \
1     * Cuomo "American Crisis" * An ethics committe...           [0, 219]   
3     * Cuomo "American Crisis" * An ethics committe...           [0, 219]   
4     * Cuomo "American Crisis" * An ethics committe...           [0, 219]   
5     @GovTimWalz @PennyWheelerMD How about making C...          [28, 124]   
7     How does #COVID19 affect us?\n\nMy partner's d...           [0, 264]   
...                                                 ...                ...   
975   96 Conservative MPs voting against the governm...           [0, 253]   
985   Tory rebellion doesn’t stop covid passes becom...            [0, 73]   
992   Boris Johnson suffers rebellion by almost 100 ...           [0, 282]   
1007  @DRTomlinsonEP @DrAnneMurphy @CareQualityComm ...         [140, 328]   
1009  @LukeHurrenpfc I thought I was replying to a g...          [15, 293]   

      ...                                               user  \
1     ...  {'id': 2582876443, 'id_str

In [251]:
#create subset with relevant colums for creating choropeth map 
finaldf2_subset = finaldf2[['actual', 'Polarity Score', 'Sentiment', 'id']]

In [252]:
finaldf2_subset["id"].value_counts()

USA    310
CAN    129
GBR    127
Name: id, dtype: int64

In [254]:
#save file to be imported back for choropeth map 
finaldf2_subset.to_excel("dfsubset2.xlsx") 